### ISSUE TITLE
Port code to c++ without OpenFramework

### ISSUE THREAD

**Issue by geotyper** [action-script/ofxWFC3D#3]
Port your  code to c++ without OpenFramework and try to start simple example.
have next picture
![Screenshot at 2020-04-19 18-22-33](https://user-images.githubusercontent.com/30755433/79692003-4a6d2580-826b-11ea-8e38-0009db92abea.png)
Think that have some errors in code because to many open lines or it normal?
Can you share picture of first example to compare results.

**Comment by geotyper** [comment]
That is look when "only turns" option choose
![Screenshot at 2020-04-19 19-00-05](https://user-images.githubusercontent.com/30755433/79692873-3f68c400-8270-11ea-81f4-0ace465eeb1f.png)

**Comment by action-script** [comment]
By the rules and models of the basic example, those results are not not normal.

![WFC_basic_02](https://user-images.githubusercontent.com/5758550/79693862-65d93000-826d-11ea-86dc-a1fbdfcc709f.png)
![WFC_basic_01](https://user-images.githubusercontent.com/5758550/79693882-7c7f8700-826d-11ea-998e-5392d6a55311.png)

Check the axis orientation of your 3D system. At OF Y goes +UP, X goes +RIGHT and Z goes -FORWARD. Could happen that the 3D assets have y and forward  wrong oriented for your implementation.

**Comment by geotyper** [comment]
Thanks for prompt answer, make procedural 3D assets, try to save orientation as in your example
Use BGFX for render, looks at orientation again.
And you dont use this parameters in XML file:
auto xmla_minheight = xmln_tile.attribute("min-height");--->0
auto xmla_maxheight = xmln_tile.attribute("max-height");----> max.z

**Comment by geotyper** [comment]
![Screenshot at 2020-04-19 20-27-10](https://user-images.githubusercontent.com/30755433/79694904-3382ff00-827c-11ea-8a50-b52c49377215.png)
Thats is right orientation?

**Comment by geotyper** [comment]
And another question, about parent node
world_node.position=Vec3(1.f, 0.f, 0.f); this coord of parent node

```
 tile_node.setParent(parent_node);
 tile_node.position=Vec3(x*grid_size.x, y*grid_size.y, z*grid_size.z);
 tile_node.rotateDegree=ofToInt(tile_cardinality[1])*90.0f;
```
when we rotate,  it rotated around center of mesh or center plus coord parent node?

**Comment by geotyper** [comment]
axis z-forward direction to screen? 
on picture from screen Z
![Screenshot at 2020-04-19 22-04-25](https://user-images.githubusercontent.com/30755433/79697173-e9a11580-8289-11ea-8486-8849d5fd9918.png)

**Comment by action-script** [comment]
> Thanks for prompt answer, make procedural 3D assets, try to save orientation as in your example
> Use BGFX for render, looks at orientation again.
> And you dont use this parameters in XML file:
> auto xmla_minheight = xmln_tile.attribute("min-height");--->0
> auto xmla_maxheight = xmln_tile.attribute("max-height");----> max.z

min-height and max-height properties are processed and used on the code. The basic example does not utilize those features. However, the advance example uses the properties.



> And another question, about parent node
> world_node.position=Vec3(1.f, 0.f, 0.f); this coord of parent node
> 
> ```
>  tile_node.setParent(parent_node);
>  tile_node.position=Vec3(x*grid_size.x, y*grid_size.y, z*grid_size.z);
>  tile_node.rotateDegree=ofToInt(tile_cardinality[1])*90.0f;
> ```
> 
> when we rotate, it rotated around center of mesh or center plus coord parent node?

The line <tile_node.rotateDeg(ofToInt( tile_cardinality[1])*90.0f, axis_y );> : rotates the tile around the model's origin. Rotations applied to the parent node will take the parent node origin as transformation point. On render, you can compound the matrices of the nested nodes like a transformation tree. 

> ![Screenshot at 2020-04-19 20-27-10](https://user-images.githubusercontent.com/30755433/79694904-3382ff00-827c-11ea-8a50-b52c49377215.png)
> Thats is right orientation?


On OF 2D screen space, the coordinate origin is top left and Y is oriented downwards to positive.
As 3D OFNode, the center of coordinates is 0,0 (center of screen) and the Y axis orientation is UP.
In both cases, Z goes positive towards the camera and negative forward.

But this does not matter. It all depends on the coorinate system and axis orientation you use on your render logic and the 3D models. Take the simetry as 2D reference and match the 3D models with that orientation.

**Comment by geotyper** [comment]
Thanks for your support, change degree of rotation to minus degree, and all look better )
![Screenshot at 2020-04-20 08-41-04](https://user-images.githubusercontent.com/30755433/79717627-d8d0be00-82e2-11ea-8816-2c53e46583f3.png)

**Comment by geotyper** [comment]
And If use cross need to add it item and neighbors to XLM? code have this simmetry
![Screenshot at 2020-04-19 20-07-59](https://user-images.githubusercontent.com/30755433/79717701-0fa6d400-82e3-11ea-80bb-2cd00efeb6ed.png)
 to XLM?

**Comment by geotyper** [comment]
if (tile_name == ground_name) ground_id = num_patterns;
if (tile_name == surround_name) surround_id = num_patterns;
When give some names to ground or surround algorithm return -1/ or nothing 
How to close borders as in your screen short?

**Comment by action-script** [comment]
> And If use cross need to add it item and neighbors to XLM? code have this simmetry
> ![Screenshot at 2020-04-19 20-07-59](https://user-images.githubusercontent.com/30755433/79717701-0fa6d400-82e3-11ea-80bb-2cd00efeb6ed.png)
> to XLM?

Yes, every tile has to be declared in the XML. It does needs a symmetry and neighbors that defined the rules for the WFC.
At the ofxWFC3D, I implemented a new kind of symmetry (called "+"). The symmetry has nothing to be with the shape of the 3D model but the relationship with the neighbors, rotations and the location possibilities.

> if (tile_name == ground_name) ground_id = num_patterns;
> if (tile_name == surround_name) surround_id = num_patterns;
> When give some names to ground or surround algorithm return -1/ or nothing
> How to close borders as in your screen short?

**ground_name** will use such tile as _gorund_ and will then force the neihbord tiles to be related to that _ground_ rules.
**surround_name** will place a tile surrounding the 4 walls of the grid and will force the WFC to calculate the result based on that rules.
Because the XML rules of the basic example are set to not connect an open pipe into a empty tile, by surrounding the grid with empty tiles, it will force the WFC to find alternatives where the borders of the pipes are closed and connected to each other.

If the WFC returns null or negative, does it mean has not found a solution. It may be due conflits on the XML rules or at the initialization parameters. If using a ground tile, make sure such tile rules match the setup and other tiles can be placed on top of such tile while not blocking other neighbors.

Imagine trying to solve a sudoku when there is already placed conflicting numbers that makes the sudoku being impossible to be solved.

**Comment by geotyper** [comment]
Thanks again for detailed answers, think that in simple example all works
![Screenshot at 2020-04-20 21-04-09](https://user-images.githubusercontent.com/30755433/79788088-ba091080-8350-11ea-92ca-bf1e7fad4055.png)

> At the ofxWFC3D, I implemented a new kind of symmetry (called "+"). The symmetry has nothing to be with the shape of the 3D model but the relationship with the neighbors, rotations and the location possibilities.

But what about symmetry (called "+"). It can be related with shape or as you write only for neighbors?

**Comment by geotyper** [comment]
and one, more:
`float bound_width=10, bound_height=22, bound_length=10;`
if bound height greater than width,length algorithm lose it possibility to find good result, try with different tile types and dimensions (5,10,5 for example)

**Comment by action-script** [comment]
> But what about symmetry (called "+"). It can be related with shape or as you write only for neighbors?

Tiles symmetry:
https://www.dropbox.com/s/zeiat1w8zre9ro8/Knots%20breakdown.png?dl=0
The "+ symmetry" is an special case that actually only use rotations and no symmetry.

**Comment by action-script** [comment]
> and one, more:
> `float bound_width=10, bound_height=22, bound_length=10;`
> if bound height greater than width,length algorithm lose it possibility to find good result, try with different tile types and dimensions (5,10,5 for example)

You are right, that was a bug.
It has been fixed on 2042eef
